docs: :memo: split guide into creating vs managing a Data Package (#1592)
seedcase-project/seedcase-sprout: docs/guide/manage-packages.qmd

Author: Given the new section titles from https://github.com/seedcase-project/seedcase-sprout/pull/1592/files#r2524037323, what do you think of these names:

- `data-package.md`
- `package-metadata.md`
- `resource-metadata.md`

---

> title: "Managing project-level metadata"

Reviewer: ```suggestion
title: "Managing package-level metadata"
```

Author: I can make this change everywhere, but I wonder if "project" is more descriptive of what this level of metadata describes? I see that the naming of the property classes are "package" and "resource" though so I can see that it makes sense to align with them.

Reviewer: Yea, since we refer to project as the Python project (`pyproject.toml`) and the repo itself. But when it comes to the metadata, it would be the Data Package.

order: 2
jupyter: python3
---

## Understanding the template for `package_properties.py`

In the last section, we saw how to write a minimal set of metadata properties
to `datapackage.json`. Here, we will take a closer look at the full
`package_properties.py` that is created by `create_properties_script()`. First,
delete file `scripts/package_properties.py` and change your `main.py` back to
look like this:

```{python}
#| filename: "main.py"
import seedcase_sprout as sp

def main():
    # Create the properties script in the default location.
    sp.create_properties_script()

if __name__ == "__main__":
    main()
```

Now run the script from terminal:

``` {.bash filename="Terminal"}
uv run main.py
```

If you now view the created `package_properties.py` script, you will see that
it lists many of the most commonly used metadata names together with comments
indicating which are required and which are optional:

<!-- Create the script where quarto can find it for building the docs -->

> {{< include _python-minimal-package-setup.qmd >}}

Reviewer: It looks like this file wasn't formatted via the VSCode "to visual mode and back again". Can you do that after adding these changes?

Author: I will do this last just so that the re-formatting doesn't remove/resolve all other comments

#| filename: "scripts/package_properties.py"
print(
    '```python',
    package_path.properties_script().read_text(),
    '```',
    sep='\n'
)
```

As you can see, there are quite a lot of metadata properties available for use.
However, as we saw in the previous section and as you can see from the comments
in this template script, the *required* metadata are not as many, so it is easy
to get started creating a Data Package and then add more metadata later as it
becomes available.

Sometimes it might feel tedious to fill out metadata properties at all and you
might be tempted to skip creating a Data Package for your data. Then it is
important to remember just how important these metadata actually are. Without
them, your data are simply a collection of files without any context or
meaning. The metadata (properties) are **crucially important** for
understanding and actually using the data in your data package!

## Creating a more complex `datapackage.json` file

Since metadata is so important, Sprout encourages users to include it by
facilitating its management via the use of Python classes as we saw in the
previous section. In the example above, you can see a couple of additional

> classes `ContributorProperties` and `SourceProperties`. Let`s create a slightly

Reviewer: ```suggestion
classes `ContributorProperties` and `SourceProperties`. Let's create a slightly
```

    ],
    ## Autogenerated:
    id="8f301286-2327-45bf-bbc8-09696d059499",
    version="0.1.0",
    created="2025-11-07T11:12:56+01:00",
)
```

You can see that we included a more involved description of the package using
the helper function `dedent()` and that we used the `ContributorProperties`
class twice as we set the `contributors` parameter to a list of two
contributors who co-created this fictional project.

Now you can edit your `main.py` file to again include the `write_properties()`
function:

```{python}
#| eval: false
#| filename: "main.py"
import seedcase_sprout as sp
from scripts.package_properties import package_properties

def main():
    # Create the metadata properties script in default location.
    sp.create_properties_script()
    # Write metadata properties from properties script to `datapackage.json`.
    sp.write_properties(properties=package_properties)

if __name__ == "__main__":
    main()
```

```{python}
#| include: false
# Only to check that it runs.
sp.write_properties(
    properties=package_properties,
    path=package_path.properties()
)
```

Then, use uv to run the script from the Terminal.

``` {.bash filename="Terminal"}
uv run main.py
```

> When you inspect the created `datapackage.json` file, you will see that it

Reviewer: ```suggestion
When you inspect the created `datapackage.json` file, you'll see that it
```

> contains all the metadata specified in `scripts/package_properties.py`:

Reviewer: ```suggestion
contains all the metadata from the `scripts/package_properties.py`:
```

```{python}
#| echo: false
#| output: asis
#| filename: datapackage.json
print(
    '```json',
    (package_path.path / 'datapackage.json').read_text(),
    '```',
    sep='\n'
)
```

If you made a mistake and want to update the properties in the current
`datapackage.json`, remember that you never need to edit the JSON file
directly. Instead, you edit the `scripts/package_properties.py` and then run
the `main.py` script to regenerate `datapackage.json`.

> ## Creating a README of the metadata properties

Reviewer: This could actually be split into its own guide! Since it doesn't make sense here.

---
title: "Package metadata"
order: 2
jupyter: python3
---

In the [previous guide](/docs/guide/packages.qmd), we created a Data Package
and we saw how to write a minimal set of metadata properties to
`datapackage.json`. Here, we'll take a closer look at the full
`package_properties.py` that's created by `create_properties_script()`.

::: callout-important

Before we get started with this section, it's necessary to delete any
previously created `scripts/package_properties.py` since
`create_properties_script()` does not overwrite this file if it already exists.

:::

First, change your `main.py` to look like this:

```{python}
#| filename: "main.py"
import seedcase_sprout as sp

def main():
    # Create the properties script in the default location.
    sp.create_properties_script()

if __name__ == "__main__":
    main()
```

Now run the script from terminal:

``` {.bash filename="Terminal"}
uv run main.py
```

> If you now view the created `package_properties.py` script, you will see that

Reviewer: ```suggestion
If you now view the created `package_properties.py` script, you'll see that
```

it includes a template containing many of the most commonly used metadata names
together with comments indicating which are required and which are optional:

<!-- Create the script where quarto can find it for building the docs -->
{{< include _python-minimal-package-setup.qmd >}}
```{python}
#| include: false
sp.create_properties_script(package_path.root())
```

```{python}
#| echo: false
#| output: asis
#| filename: "scripts/package_properties.py"
print(
    '```python',
    package_path.properties_script().read_text(),
    '```',
    sep='\n'
)
```

As you can see, there are a lot of available metadata properties.
However, as we saw in the previous section and as you can see from the comments
in this template script, there aren't that many *required* metadata, so it is easy

> to get started creating a Data Package and then add more metadata later as it

Reviewer: ```suggestion
get started creating a Data Package and add more metadata later as
```

becomes available.

Sometimes it might feel tedious to fill out metadata properties at all and you
might be tempted to skip creating a Data Package for your data. Then it is

> important to remember just how important these metadata actually are. Without

Reviewer: ```suggestion
important to remember just how vital these metadata actually are. Without
```

to `datapackage.json`. Here, we will take a closer look at the full
`package_properties.py` that is created by `create_properties_script()`. First,
delete file `scripts/package_properties.py` and change your `main.py` back to
look like this:

```{python}
#| filename: "main.py"
import seedcase_sprout as sp

def main():
    # Create the properties script in the default location.
    sp.create_properties_script()

if __name__ == "__main__":
    main()
```

Now run the script from terminal:

``` {.bash filename="Terminal"}
uv run main.py
```

If you now view the created `package_properties.py` script, you will see that
it lists many of the most commonly used metadata names together with comments
indicating which are required and which are optional:

<!-- Create the script where quarto can find it for building the docs -->
{{< include _python-minimal-package-setup.qmd >}}
```{python}
#| include: false
sp.create_properties_script(package_path.root())
```

```{python}
#| echo: false
#| output: asis
#| filename: "scripts/package_properties.py"
print(
    '```python',
    package_path.properties_script().read_text(),
    '```',
    sep='\n'
)

> ```

Author: Before:

<img width="582" height="884" alt="Image" src="https://github.com/user-attachments/assets/d8b00973-bbd2-4790-b25d-ac42e78e0289" />

After:

<img width="582" height="884" alt="Image" src="https://github.com/user-attachments/assets/578f4f10-c3a6-4f99-983f-7df9c96fb71c" />
